cpp/src/io/text/multibyte_split.cu

```diff
@@ -0,0 +1,543 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/io/text/data_chunk_source.hpp>
+#include <cudf/io/text/multistate.hpp>
+#include <cudf/io/text/trie.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_pool.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_scan.cuh>
+#include <cub/warp/warp_reduce.cuh>
+
+#include <iostream>
+#include <memory>
+
+namespace {
+
+template <typename Dividend, typename Divisor>
+inline constexpr auto ceil_div(Dividend dividend, Divisor divisor)
+{
+  return dividend / divisor + (dividend % divisor != 0);
+}
+
+using multistate = cudf::io::text::multistate;
+
+enum class scan_tile_status : uint8_t {
+  oob,
+  invalid,
+  partial,
+  inclusive,
+};
+
+template <typename T>
+struct scan_tile_state_view {
+  uint64_t num_tiles;
+  scan_tile_status* tile_status;
+  T* tile_partial;
+  T* tile_inclusive;
+
+  __device__ inline void initialize_status(cudf::size_type base_tile_idx,
+                                           cudf::size_type count,
+                                           scan_tile_status status)
+  {
+    auto thread_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (thread_idx < count) {  //
+      // this is probably UB without taking in to account tile_states being assigned multiple ties
+      // due to modulo operator
+      tile_status[(base_tile_idx + thread_idx) % num_tiles] = status;
+    }
+  }
+
+  __device__ inline void set_partial_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_partial + offset, value);
+    __threadfence();
```

Comment:
Using a `__threadfence()` directly is almost certainly the wrong thing to do in lieu of using an  appropriate `cuda::atomic` object.

Reply:
Most definitely. I could use your help improving this code in particular.

Reply:
It appears that the `thread_fence` here may not be required at all.

Reply:
I think `thread_fence` is needed to make sure the writes of `value` to the respective `tile_partial` is observed *before* the write to `tile_status` (i.e., signalling that the `value` is there). Otherwise, it could be that a thread is reading the state `scan_tile_status::partial`, hence, assuming it can safely read the tile's partial, while the write of `value` may not be observed yet. 
But the discussion probably just emphasizes the point that we need `cuda::atomic` :-)

Reply:
Updated to use `cudf::atomic` for `tile_state` but left `tile_partial` and `tile_inclusive` alone, but according to cuda docs, the instructions issued by `cub::ThreadStore<cub::STORE_CG>` just _hints_ at cache invalidation, but doesn't indicate the scope of the memory operation, so the others may need to be updated to use atomic as well.